audio/filter/af_scaletempo2.c

```diff
@@ -112,6 +112,17 @@ static void process(struct mp_filter *f)
             double frame_delay = mp_scaletempo2_get_latency(&p->data, p->speed)
                 + out_samples * p->speed;
             mp_aframe_set_pts(out, pts - frame_delay / mp_aframe_get_effective_rate(out));
+
+            if (p->sent_final) {
+                double remain_pts = pts - mp_aframe_get_pts(out);
+                double rate = mp_aframe_get_effective_rate(out) / p->speed;
+                int max_samples = MPMAX(0, (int) (remain_pts * rate));
+                // truncate final packet to expected length
+                if (out_samples >= max_samples) {
+                    out_samples = max_samples;
+                    mp_scaletempo2_reset(&p->data);
```

Comment:
Why is that reset necessary?

Reply:
Once the `mp_scaletempo2_set_final` method is called, the internals may produce more audio than expected, depending on the target block position and which blocks are chosen if WSOLA runs. The reset clears all this internal state to guarantees that `mp_scaletempo2_frames_available` returns false again.

After that, processing can continue as normal again. E.g. if there's a format change, then the first `while` loop of `process()` will reinitialize the internals with the new format (first iteration sets `p->initialized` to false and the next loop calls `init_scaletempo2`)

Reply:
Thanks, that makes more sense now. A comment explaining that would be good.